program/rust/src/time_machine_types.rs

```diff
@@ -6,20 +6,375 @@ use crate::c_oracle_header::{
     PC_PRICE_T_COMP_OFFSET,
 };
 use crate::error::OracleError;
+use crate::utils::try_convert;
 use bytemuck::{
     Pod,
     Zeroable,
 };
-use solana_program::msg;
+use std::cmp;
 
+//To make it easy to change the types to allow for more usefull
+//running sums if needed
+use {
+    i64 as SignedTrackerRunningSum,
+    u64 as UnsignedTrackerRunningSum,
+};
+//multiply by this to make up for errors that result when dividing integers
+pub const SMA_TRACKER_PRECISION_MULTIPLIER: i64 = 10;
+
+
+pub trait Tracker<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64> {
+    ///initializes a zero initialized tracker
+    fn initialize(&mut self) -> Result<(), OracleError>;
+
+    ///add a new price to the tracker
```

Comment:
1.  Why we need to pass the previous price is because of the formula of SMA tracking! once you migrate everything into one struct you won't need it, but I see no point in storing stuff that are in price_t again in each tracker. Though we can do it if you think that's is significantly cleaner